### PR TITLE
Panel UI: Numeric filtering

### PR DESCRIPTION
**Pull request by gtk-grafana**
Adds numeric filtering to duration, bytes, and float fields.

Fixes: https://github.com/grafana/explore-logs/issues/891

image
image
image

Moves the include/exclude filters on sparse fields/labels to a new button group which combines with numeric filtering, for applicable panels.
image
image

### CODE REVIEW

File: src/Components/IndexScene/IndexScene.tsx [hunk 1]
```diff
@@ -180,6 +181,18 @@ export class IndexScene extends SceneObjectBase<IndexSceneState> {
     this._subs.add(timeRange.subscribeToState(this.limitMaxInterval(timeRange)));
   }
 
+  private setVariableOperators() {
+    const fieldsVar = getFieldsVariable(this);
+    // No wip set when this is called, so once the operations must be consistent for all filters
+    // @todo we'll need to split out the duration, bytes, and float filters out of the VAR_FIELDS variable and into a new variable if we want to keep comparison operations out of string, int, etc filters
```
Comment (gtk-grafana) [inline src/Components/IndexScene/IndexScene.tsx:187]: remove todo, create issue

File: src/Components/ServiceScene/Breakdowns/FieldsAggregatedBreakdownScene.tsx [hunk 2]
```diff
@@ -44,59 +53,77 @@ export class FieldsAggregatedBreakdownScene extends SceneObjectBase<FieldsAggreg
...
+      // populate set of new list of fields
+      const newFieldsSet = new Set<string>(newNamesField?.values);
+      const updatedChildren = layout.state.children as SceneCSSGridItem[];
+
+      // Iterate through all the existing panels
+      for (let i = 0; i < updatedChildren.length; i++) {
+        const gridItem = layout.state.children[i] as SceneCSSGridItem;
+        const panel = gridItem.state.body as VizPanel;
+
+        if (newParser) {
```
Comment (gtk-grafana) [inline src/Components/ServiceScene/Breakdowns/FieldsAggregatedBreakdownScene.tsx:81]: This fixes a bug in the field names scene where adding a field filter with a different parser then the panel requires i.e. a metadata field panel breaks when a logfmt/json field include/exclude/numeric filter is added.

File: src/Components/ServiceScene/GoToExploreButton.tsx [hunk 3]
```diff
@@ -22,10 +23,14 @@ export const GoToExploreButton = ({ exploration }: GoToExploreButtonState) => {
     const datasource = getDataSource(exploration);
     const expr = getQueryExpr(exploration).replace(/\s+/g, ' ').trimEnd();
     const timeRange = sceneGraph.getTimeRange(exploration).state.value;
+    const displayedFields = getDisplayedFields(exploration);
```
Comment (gtk-grafana) [inline src/Components/ServiceScene/GoToExploreButton.tsx:26]: Why are these changes in this PR?

File: src/Components/ServiceScene/LogsPanelScene.tsx [hunk 4]
```diff
@@ -130,13 +130,14 @@ export class LogsPanelScene extends SceneObjectBase<LogsPanelSceneState> {
...
     if (row?.rowId && this.state.body) {
+      const parent = this.getParentScene();
       const timeRange = sceneGraph.getTimeRange(this.state.body);
       const buttonRef = event.currentTarget instanceof HTMLButtonElement ? event.currentTarget : undefined;
       copyText(
         generateLogShortlink(
           'panelState',
           {
-            logs: { id: row.uid },
+            logs: { id: row.uid, displayedFields: parent.state.displayedFields },
```
Comment (gtk-grafana) [inline src/Components/ServiceScene/LogsPanelScene.tsx:140]: Why are these changes in this PR

File: src/services/text.ts [hunk 5]
```diff
@@ -33,7 +33,17 @@ export enum UrlParameterType {
   To = 'to',
 }
 
-type PermalinkDataType = Record<string, string | number | Record<string, string | number>>;
+type PermalinkDataType =
```
Comment (gtk-grafana) [inline src/services/text.ts:36]: why are these changes in this PR?

File: tests/fixtures/explore.ts [hunk 6]
```diff
@@ -222,3 +222,11 @@ export class ExplorePage {
     });
   }
 }
+
+export async function compareLocators(firstLocator: Locator, secondLocator: Locator): Promise<boolean> {
```
Comment (gtk-grafana) [inline tests/fixtures/explore.ts:226]: unused remove

File: generator/generator.go [hunk 7]
```diff
@@ -144,7 +144,7 @@ var generators = map[model.LabelValue]map[model.LabelValue]LogGenerator{
 var noisyTempo = func(ctx context.Context, logger *AppLogger, metadata push.LabelsAdapter) {
 	const fmt1 = `level=debug ts=%s caller=broadcast.go:48 msg="Invalidating forwarded broadcast" key=collectors/compactor version=%d oldVersion=%d content=[compactor-%s] oldContent=[compactor-%s]`
 	const fmt2 = `level=warn ts=%s caller=instance.go:43 msg="TRACE_TOO_LARGE: max size of trace (52428800) exceeded tenant %s"`
-	const fmt3 = `level=info ts=%s caller=compactor.go:242 msg="flushed to block" bytes=%d objects=%d values=%d`
+	const fmt3 = `level=info ts=%s caller=compactor.go:242 msg="flushed to block" bytes=%dB objects=%d values=%d`
```
Comment (gtk-grafana) [inline generator/generator.go:147]: the unit was missing so Loki was interpreting this as an int instead of a bytes field

**Comment by matyax** [comment]
Do we have control over this padding? Removing it and just keeping theme.spacing(8) seems to better align stuff and have less white space.

Padding

vs

imagen

It also aligns the form controls.

File: src/Components/ServiceScene/Breakdowns/NumericFilterPopoverScene.tsx [hunk 8]
```diff
@@ -0,0 +1,508 @@
...
+      units = { ltu: byteUnitValues.B, gtu: byteUnitValues.B };
+    } else if (fieldType === 'duration') {
+      units = { ltu: durationUnitValues.s, gtu: durationUnitValues.s };
+    } else if (fieldType === 'float') {
+      units = { ltu: '', gtu: '' };
+    } else {
+      throw new Error(`field type incorrectly defined: ${fieldType}`);
+    }
+
+    // @todo - how to avoid type assertion?
```
Comment (matyax) [inline src/Components/ServiceScene/Breakdowns/NumericFilterPopoverScene.tsx:94]: Type guards.

File: src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx [hunk 9]
```diff
@@ -73,6 +190,36 @@ export class SelectLabelActionScene extends SceneObjectBase<SelectLabelActionSce
             Select
           </Button>
         )}
+
+        {popover && (
+          <PopoverController content={<popover.Component model={popover} />}>
+            {(showPopper, hidePopper, popperProps) => {
```
Comment (matyax) [inline src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx:196]: Popover?

File: src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx [hunk 10]
```diff
@@ -110,6 +261,23 @@ export class SelectLabelActionScene extends SceneObjectBase<SelectLabelActionSce
     );
   }
 
+  public onClickNumericFilter = (variableType: VariableFilterType) => {
+    const detectedFieldFrame = getDetectedFieldsFrame(this);
+    const fieldType = getDetectedFieldType(this.state.labelName, detectedFieldFrame);
+
+    // @todo, what can we do if we can't determine the field type?
```
Comment (matyax) [inline src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx:268]: What are the chances of this happening?

**Review by matyax — COMMENTED** [review]
Love this feature, great work! Left a few comments and suggestions around.

File: src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx (hunk 9 above)
Comment (gtk-grafana) [reply src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx:196]: https://developers.grafana.com/ui/latest/index.html?path=/story/pickers-and-editors-colorpicker-popovers--basic
It's a grafana UI element, but I think I'm misunderstanding the question

File: src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx (hunk 10 above)
Comment (gtk-grafana) [reply src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx:268]: I would hope very low, should only happen with an unsupported or broken Loki that isn't returning the field types in the detected_field response. I believe my concern was that the detected_fields response isn't there, but if it's not there we wouldn't have any field panels to render this component on 🤷

File: src/Components/ServiceScene/Breakdowns/NumericFilterPopoverScene.tsx (hunk 8 above)
Comment (gtk-grafana) [reply src/Components/ServiceScene/Breakdowns/NumericFilterPopoverScene.tsx:94]: Easier said then done, if you have some time to take a look at this we can discuss how to work around this, if we had a hard requirement to not assert here I would likely remove the discriminated union on the fieldType and just make it a string, which might be more likely to introduce bugs in the future then the type assertion.

File: src/Components/ServiceScene/Breakdowns/FieldsAggregatedBreakdownScene.tsx [hunk 11]
```diff
@@ -44,59 +53,77 @@ export class FieldsAggregatedBreakdownScene extends SceneObjectBase<FieldsAggreg
...
+  };
+
+  private updateChildren(newState: QueryRunnerState, newParser: ParserType | undefined = undefined) {
+    const detectedFieldsFrame = getDetectedFieldsFrameFromQueryRunnerState(newState);
+    const newNamesField = getDetectedFieldsNamesFromQueryRunnerState(newState);
+    const newParsersField = getDetectedFieldsParsersFromQueryRunnerState(newState);
+    const cardinalityMap = this.calculateCardinalityMap(newState);
+
+    // Iterate through all the layouts
+    this.state.body?.state.layouts.forEach((layoutObj) => {
```
[the remark this answers is not in the document]
Comment (gtk-grafana) [reply src/Components/ServiceScene/Breakdowns/FieldsAggregatedBreakdownScene.tsx]: I was able to remove all but one assertion, there is similar code in the labels scene as well we should probably clean up at some point.

Also we might want to do a bigger refactor to the layoutSwitcher and see if we can solve this with templating or maybe splitting that class apart

File: src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx (hunk 9 above)
Comment (matyax) [reply src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx:196]: Unexpected! Thanks for the info.

File: src/Components/IndexScene/IndexScene.tsx (hunk 1 above)
Comment (matyax) [reply src/Components/IndexScene/IndexScene.tsx:187]: Friendly reminder to remove comment and add the issue.

**Review by matyax — APPROVED** [review]
🚢 🇮🇹

**Pull request merged**